qutip/cy/stochastic.pyx

```diff
@@ -33,127 +33,2199 @@
 
-    psi1 = spmv_csr(A[0].data, A[0].indices, A[0].indptr, psi)
-    n1 = np.linalg.norm(psi1)
-    if n1 != 0:
-        return psi1 / n1 - psi
-    else:
-        return - psi
+@cython.boundscheck(False)
+cdef void _zero_4d(complex[:,:,:,::1] x):
+    """ x *= 0 """
+    cdef int l = x.shape[0]*x.shape[1]*x.shape[2]*x.shape[3]
+    zdscal(&l, &DZERO, <complex*>&x[0,0,0,0], &ONE)
 
+"""functions for ensuring that the states stays physical"""
+@cython.cdivision(True)
+@cython.boundscheck(False)
+cdef void _normalize_inplace(complex[::1] vec):
+    """ make norm of vec equal to 1"""
+    cdef int l = vec.shape[0]
+    cdef double norm = 1.0/_dznrm2(vec)
+    zdscal(&l, &norm, <complex*>&vec[0], &ONE)
 
+@cython.cdivision(True)
 @cython.boundscheck(False)
-@cython.wraparound(False)
-cpdef d1_psi_homodyne(t, psi, A, args):
+cdef void _normalize_rho(complex[::1] rho):
+    """ Ensure that the density matrix trace is one and
+    that the composing states are normalized.
     """
-    Cython version of d2_psi_homodyne. See d2_psi_homodyne for docs.
-    """
-    e1 = cy_expect_psi_csr(A[1].data, A[1].indices, A[1].indptr, psi, 0)
-    return 0.5 * (e1 * spmv_csr(A[0].data, A[0].indices, A[0].indptr, psi) -
-                  spmv_csr(A[3].data, A[3].indices, A[3].indptr, psi) -
-                  0.25 * e1 ** 2 * psi)
+    cdef int l = rho.shape[0]
+    cdef int N = np.sqrt(l)
+    cdef complex[::1,:] mat = np.reshape(rho, (N,N), order="F")
+    cdef complex[::1,:] eivec = np.zeros((N,N), dtype=complex, order="F")
+    cdef double[::1] eival = np.zeros(N)
+    ZHEEVR(mat, &eival[0], eivec, N)
+    _zero(rho)
+    cdef int i, j, k
+    cdef double sum
 
+    sum = 0.
+    for i in range(N):
+        _normalize_inplace(eivec[:,i])
+        if eival[i] < 0:
+            eival[i] = 0.
+        sum += eival[i]
+    if sum != 1.:
+        for i in range(N):
+            eival[i] /= sum
 
-@cython.boundscheck(False)
-@cython.wraparound(False)
-cpdef d2_psi_homodyne(t, psi, A, args):
-    """
-    Cython version of d2_psi_homodyne. See d2_psi_homodyne for docs.
-    """
-    e1 = cy_expect_psi_csr(A[1].data, A[1].indices, A[1].indptr, psi, 0)
-    return [spmv_csr(A[0].data, A[0].indices, A[0].indptr, psi) - 0.5 * e1 * psi]
+    for i in range(N):
+        for j in range(N):
+            for k in range(N):
+                rho[j+N*k] += conj(eivec[k,i])*eivec[j,i]*eival[i]
 
-# -----------------------------------------------------------------------------
-# density matric (vector form) d1 and d2 functions
-#
 
-@cython.boundscheck(False)
-@cython.wraparound(False)
-cpdef cy_d1_rho_photocurrent(double t, cnp.ndarray[CTYPE_t, ndim=1] rho_vec,
-                             object A, object args):
-    """
-    Photo-current D1 function
-    """
-    n = A[4] + A[5]
-    e1 = cy_expect_rho_vec_csr(n.data, n.indices, n.indptr, rho_vec, 0)
-    return 0.5 * (e1 * rho_vec - spmv_csr(n.data, n.indices, n.indptr, rho_vec))
+"""
+Solver:
```

Comment:
I made them enum. Now you have:
`if self.solver is EULER_SOLVER:`
which is a lot more readable.
I used capitals since enum are constant, even thought it is heavy.